### ISSUE TITLE
Depending on a Git repository with a git submodule throws an NPE

### ISSUE THREAD

**Issue by Furetur** [gradle/gradle#18013]
I reproduce the issue with 2 simple repositories: a [library](https://github.com/Furetur/simplelib) and an [application](https://github.com/Furetur/this-project-breaks-gradle) that depends on that _library_.

* I'm trying to add the _library_ that is a Git repository as a dependency to my _application_ with

```kotlin
// settings.gradle.kts

val myRepo = "https://github.com/Furetur/simplelib.git"
val myRepoProjectName = "me.user"
sourceControl {
    gitRepository(java.net.URI.create(myRepo)) {
        producesModule("$myRepoProjectName:simplelib")
    }
}

// build.gradle.kts

dependencies {
    testImplementation(kotlin("test"))

    val myRepoProjectName = "me.user"
    implementation("$myRepoProjectName:simplelib") {
        version {
            branch = "master-with-submodule"
        }
    }
}
```
* The _library_ has 1 git submodule. The submodule does not contain any code, only a README.
* Gradle fails to build the _application_. However, if I remove the submodule from the _library_, the build passes.

### Expected Behavior

Gradle should build the _application_ if it depends on a _library_ that has a Git submodule.

### Current Behavior

Gradle throws a NullPointerException: [stacktrace.txt](https://github.com/gradle/gradle/files/6975169/stacktrace.txt)

### Context

I have a library that has tests itself on mock data. That mock data is located in another Git repo and is attached to the library as a git submodule. When I'm trying to depend on that library Gradle fails to build the application. 

### Steps to Reproduce 

[Library](https://github.com/Furetur/simplelib) and [application](https://github.com/Furetur/this-project-breaks-gradle).

1. Clone the _application_ and run `./gradlew build`.
2. It should fail with that NPE 
3. If you change the branch of the _library_ in _build.gradle.kts_ to `master` the gradle build will pass. The only difference between branches `master` and `master-with-submodule` is that the latter has a Git submodule that contains a README.

### Your Environment

Build scan URL: https://scans.gradle.com/s/jmavmepum35is

**Comment by mlopatkin** [comment]
One interesting thing is that the first invocation of `./gradlew build` after checking out works, only consecutive ones fail.

Unfortunately, source dependencies are not in the focus so the issue is unlikely to be addressed in the short or medium term.

**Comment by jbartok** [comment]
Should be fixed by #20090
